xdebug_com.c

```diff
@@ -25,165 +25,205 @@
-	lookup_hostname(hostname, &address.sin_addr);
-	address.sin_family = AF_INET;
-	address.sin_port = htons((unsigned short)dport);
-
-	sockfd = socket(address.sin_family, SOCK_STREAM, 0);
-	if (sockfd == SOCK_ERR) {
+	
+	/* Make a string of the port number that can be used with getaddrinfo */
+	sprintf(sport, "%d", dport);
+
+	/* Create hints for getaddrinfo saying that we want IPv4 and IPv6 TCP stream sockets */
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = IPPROTO_TCP;
+	hints.ai_flags = AI_PASSIVE;
+
+	/* Call getaddrinfo and return SOCK_ERR if the call fails for some reason */
+	if ((status = getaddrinfo(hostname, sport, &hints, &remote)) != 0) {
 #ifndef DEBUGGER_FAIL_SILENTLY
 #if WIN32|WINNT
-		printf("create_debugger_socket(\"%s\", %d) socket: %d\n",
+		printf("create_debugger_socket(\"%s\", %d) getaddrinfo: %d\n",
 					hostname, dport, WSAGetLastError());
 #else
-		printf("create_debugger_socket(\"%s\", %d) socket: %s\n",
+		printf("create_debugger_socket(\"%s\", %d) getaddrinfo: %s\n",
 					hostname, dport, strerror(errno));
 #endif
 #endif
-		return -1;
+		return SOCK_ERR;
 	}
 
-	/* Put socket in non-blocking mode so we can use select for timeouts */
-	timeout.tv_sec = 0;
-	timeout.tv_usec = 200000;
+	/* Go through every returned IP address */
+	for (ptr = remote; ptr != NULL; ptr = ptr->ai_next) {
+		/* Try to create the socket. If the creation fails continue on with the next IP address in the list */
+		if ((sockfd = socket(ptr->ai_family, ptr->ai_socktype, ptr->ai_protocol)) == SOCK_ERR) {
+#ifndef DEBUGGER_FAIL_SILENTLY
+#if WIN32|WINNT
+			printf("create_debugger_socket(\"%s\", %d) socket: %d\n",
+						hostname, dport, WSAGetLastError());
+#else
+			printf("create_debugger_socket(\"%s\", %d) socket: %s\n",
+						hostname, dport, strerror(errno));
+#endif
+#endif
+			continue;
+		}
+
+		/* Put socket in non-blocking mode so we can use select for timeouts */
+		timeout.tv_sec = 0;
+		timeout.tv_usec = 200000;
 
 #ifdef WIN32
-	ioctlsocket(sockfd, FIONBIO, (u_long*)&yes);
+		ioctlsocket(sockfd, FIONBIO, (u_long*)&yes);
 #else
-	fcntl(sockfd, F_SETFL, O_NONBLOCK);
+		fcntl(sockfd, F_SETFL, O_NONBLOCK);
 #endif
 
-	/* connect */
-	status = connect(sockfd, (struct sockaddr*)&address, sizeof(address));
-	if (status < 0) {
+		/* Try to connect to the newly created socket */
+		/* Worth noting is that the port is set in the getaddrinfo call before */
+		status = connect(sockfd, ptr->ai_addr, ptr->ai_addrlen);
+		
+		/* Determine if we got a connection. If no connection could be made we close the socket and continue with the next IP address in the list */
+		if (status < 0) {
 #ifdef WIN32
-		errno = WSAGetLastError();
-		if (errno != WSAEINPROGRESS && errno != WSAEWOULDBLOCK) {
-			close(sockfd);
-			return -1;
-		}
+			errno = WSAGetLastError();
+			if (errno != WSAEINPROGRESS && errno != WSAEWOULDBLOCK) {
+#ifndef DEBUGGER_FAIL_SILENTLY
```

Comment:
If it is a nested if(n)def, I usually put a space in between the # (which must be in the first column)( and the if(n)def or endif.